File: Engine/src/main/java/org/etieskrill/engine/graphics/texture/animation/AnimatedTexturePlayer.java
package org.etieskrill.engine.graphics.texture.animation;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;

@RequiredArgsConstructor
public class AnimatedTexturePlayer {

    private final @Getter AnimatedTexture texture;
    private final Behaviour behaviour;

    private @Setter float speed = 1;
    private float time = 0;
    private @Getter boolean playing;

    public enum Behaviour {
        ONCE,
        REPEAT
    }

    public AnimatedTexturePlayer(AnimatedTexture texture) {
        this.texture = texture;
        this.behaviour = Behaviour.REPEAT;
    }

    public void play() {
        time = 0;
        playing = true;
    }

    public void resume() {
        playing = true;
    }

    public void stop() {
        playing = false;
    }

    public void update(double delta) {
        if (playing) {
            time += speed * (float) delta;
        }
    }

    public int getFrame() {
        float time;
        switch (behaviour) {
            case REPEAT -> time = this.time % texture.getMetaData().getDuration();
            case ONCE -> time = this.time;
            case null -> throw new IllegalStateException("Animation behaviour may not be null");
        }

        int numFrames = texture.getMetaData().getFrames().size();
        int index = -1;
        for (int i = 0; i < numFrames - 1; i++) {
            float previousTime = texture.getMetaData().getFrames().get(i).getTime();
            float frameTime = texture.getMetaData().getFrames().get(i + 1).getTime();

            if (frameTime >= time && previousTime <= time) {
                index = i;
                break;
            }
        }

        float lastFrameTime = texture.getMetaData().getFrames().get(numFrames - 1).getTime();
        float animationDuration = texture.getMetaData().getDuration();
        if (index == -1 && time >= lastFrameTime && time < animationDuration) {
            index = numFrames - 1;
        }

        if (index == -1) {
            if (behaviour == Behaviour.REPEAT) {
                index = 0;
            } else if (behaviour == Behaviour.ONCE) {
                index = texture.getMetaData().getFrames().size() - 1;
            }
        }

        return index;
    }

}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/texture/animation/AnimatedTexturePlayer.kt
package org.etieskrill.engine.graphics.texture.animation

enum class AnimatedTexturePlayerBehaviour { ONCE, REPEAT }

class AnimatedTexturePlayer(
    val texture: AnimatedTexture,
    private val behaviour: AnimatedTexturePlayerBehaviour = AnimatedTexturePlayerBehaviour.REPEAT
) {

    var speed = 1f
    private var time = 0f
    var playing = false

    fun play() {
        time = 0f
        playing = true
    }

    fun resume() {
        playing = true
    }

    fun stop() {
        playing = false
    }

    fun update(delta: Double) {
        if (playing) {
            time += speed * delta.toFloat()
        }
    }

    fun getFrame(): Int {
        val time = when (behaviour) {
            AnimatedTexturePlayerBehaviour.REPEAT -> time % texture.metadata.duration
            AnimatedTexturePlayerBehaviour.ONCE -> time
        }

        val frames = texture.metadata.frames

        return when {
            time <= frames[0].time -> 0
            time >= frames.last().time -> frames.lastIndex
            else -> frames.indices.drop(1)
                .indexOfFirst { i -> time in frames[i - 1].time..frames[i].time }
        }
    }

}
